plutus-core/plutus-ir/src/PlutusIR/Transform/CaseOfCase.hs

```diff
@@ -0,0 +1,148 @@
+-- editorconfig-checker-disable-file
+{-# LANGUAGE RankNTypes          #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE ViewPatterns        #-}
+module PlutusIR.Transform.CaseOfCase where
+
+import PlutusCore qualified as PLC
+import PlutusCore.Name qualified as PLC
+import PlutusIR
+import PlutusIR.Arity
+import PlutusIR.Contexts
+import PlutusIR.Transform.Rename ()
+
+import Control.Lens hiding (cons)
+import PlutusIR.Analysis.Builtins
+import PlutusIR.Analysis.VarInfo
+
+{-|
+Perform the case-of-case transformation. This pushes
+case expressions into the case branches of other case
+expressions, which can often yield optimization opportunities.
+
+Example:
+@
+    case (case s of { C1 a -> x; C2 b -> y; }) of
+      D1 -> w
+      D2 -> z
+
+    -->
+
+    case s of
+      C1 a -> case x of { D1 -> w; D2 -> z; }
+      C2 b -> case y of { D1 -> w; D2 -> z; }
```

Comment:
Any considerations regarding the fact that this duplicates branches and potentially makes the output exponentially big?